neqo-http3/src/connection.rs

```diff
@@ -303,9 +298,9 @@ pub struct Http3Connection {
     pub qpack_encoder: Rc<RefCell<QPackEncoder>>,
     pub qpack_decoder: Rc<RefCell<QPackDecoder>>,
     settings_state: Http3RemoteSettingsState,
-    streams_with_pending_data: BTreeSet<StreamId>,
-    pub send_streams: HashMap<StreamId, Box<dyn SendStream>>,
-    pub recv_streams: HashMap<StreamId, Box<dyn RecvStream>>,
+    streams_with_pending_data: HashSet<StreamId>,
+    pub send_streams: IndexMap<StreamId, Box<dyn SendStream>>,
```

Comment:
Conceptually, why is an `IndexMap` faster than a `HashMap` here? Both end up hashing the key, right?

Reply:
It's not faster, it's possibly even slower. I made the change based on this comment: https://github.com/mozilla/neqo/pull/2342/files/bd061693b40e91b846c4d4cd1bc0ecfcd27c4e45#r1945509653

Reply:
So I'm not completely sure that we need consistently ordered iteration here.  We rely on that in the transport crate because we want frame sending to follow a predictable pattern (clear the old stuff out first).  However, that's not something we rely on at this layer of the stack, so whatever is fastest might be the right answer.

Maybe the first question we have is whether we need (insertion-)ordered iteration over this collection?

Reply:
First, I think we should differentiate between RX and TX.

On RX (both here and in transport), I don't see a point in using anything that maintains order or has semantics that go beyond a simple map or set. I think all we do here is lookups based on received data.

On TX, (both here and in transport), do we **really** care that iteration is consistently ordered? All these data structures _should_ mostly have a stable order while there are no insertions or removals. When there are, do we care that we'll have a different order afterwards? I can't convince myself that we do.

In other words, I think we should use whatever is fastest.

(Also, unrelated, I find the duplication/overlap between the transport and http3 crates really odd.)